### PR TITLE
Upgrade glibcxx

### PR DESCRIPTION
**Pull request by crazyxy**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by crazyxy** [comment]
/test istio-lint-master
/test istio-racetest-master
test integ-mixer-k8s-presubmit-tests-master
/test integ-new-install-k8s-presubmit-tests-master

**Comment by crazyxy** [comment]
> we need a tagging scheme on this image before we can update it or we may break older releases.
> 
> Why do we need this here but not in the prow builder? do we just need to build this again to update and not explicitly update?

I don't think it will break older release since it's backward compatible. I don't know why prow build has higher GLIB_CXX version that istio base image. It may be used to compile WASM. Rebuild the base doesn't help.

**Comment by howardjohn** [comment]
The dockerfile is here https://github.com/istio/test-infra/blob/master/docker/prowbazel/Dockerfile

**Comment by crazyxy** [comment]
> The dockerfile is here https://github.com/istio/test-infra/blob/master/docker/prowbazel/Dockerfile

Yes. I pull the image and check the GLIBCXX version. It supports 
```
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
GLIBCXX_3.4.3
GLIBCXX_3.4.4
GLIBCXX_3.4.5
GLIBCXX_3.4.6
GLIBCXX_3.4.7
GLIBCXX_3.4.8
GLIBCXX_3.4.9
GLIBCXX_3.4.10
GLIBCXX_3.4.11
GLIBCXX_3.4.12
GLIBCXX_3.4.13
GLIBCXX_3.4.14
GLIBCXX_3.4.15
GLIBCXX_3.4.16
GLIBCXX_3.4.17
GLIBCXX_3.4.18
GLIBCXX_3.4.19
GLIBCXX_3.4.20
GLIBCXX_3.4.21
GLIBCXX_3.4.22
GLIBCXX_3.4.23
GLIBCXX_3.4.24
GLIBCXX_3.4.25
GLIBCXX_3.4.26
```

Istio base image doesn't support 

```
GLIBCXX_3.4.22
GLIBCXX_3.4.23
GLIBCXX_3.4.24
GLIBCXX_3.4.25
GLIBCXX_3.4.26

File: docker/Dockerfile.xenial_debug [hunk 1]
```diff
@@ -26,3 +26,11 @@ RUN apt-get update && \
    && apt-get upgrade -y \
    && apt-get clean \
    && rm -rf  /var/log/*log /var/lib/apt/lists/* /var/log/apt/* /var/lib/dpkg/*-old /var/cache/debconf/*-old
+
+# upgrade libstdc++6
+RUN apt-get update && \
+    apt-get install --no-install-recommends -y software-properties-common
```
Comment (PiotrSikora) [inline docker/Dockerfile.xenial_debug]: Nit: you can pull this into list above.

File: docker/Dockerfile.xenial_debug [hunk 2]
```diff
@@ -26,3 +27,11 @@ RUN apt-get update && \
    && apt-get upgrade -y \
    && apt-get clean \
    && rm -rf  /var/log/*log /var/lib/apt/lists/* /var/log/apt/* /var/lib/dpkg/*-old /var/cache/debconf/*-old
+
+# upgrade libstdc++6
+# hadolint ignore=DL3008
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test \
+ && apt-get update \
+ && apt-get install --no-install-recommends -y --only-upgrade libstdc++6 \
```
Comment (howardjohn) [inline docker/Dockerfile.xenial_debug]: prowbazel installs `libc++-8-dev` -- is it ok this is different?

File: docker/Dockerfile.xenial_debug [hunk 3]
```diff
@@ -26,3 +27,11 @@ RUN apt-get update && \
    && apt-get upgrade -y \
    && apt-get clean \
    && rm -rf  /var/log/*log /var/lib/apt/lists/* /var/log/apt/* /var/lib/dpkg/*-old /var/cache/debconf/*-old
+
+# upgrade libstdc++6
+# hadolint ignore=DL3008
```
Comment (howardjohn) [inline docker/Dockerfile.xenial_debug]: can we fix this instead of ignoring?

File: docker/Dockerfile.xenial_debug [hunk 4]
```diff
@@ -26,3 +27,11 @@ RUN apt-get update && \
...
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test \
+ && apt-get update \
```
Comment (howardjohn) [inline docker/Dockerfile.xenial_debug]: can we merge this all with the install block above?

File: docker/Dockerfile.xenial_debug (hunk 2 above)
Comment (crazyxy) [reply docker/Dockerfile.xenial_debug]: It's ok. After upgrading libstdc++6, GLIBCXX_3.4.22 is supported.

File: docker/Dockerfile.xenial_debug (hunk 3 above)
Comment (crazyxy) [reply docker/Dockerfile.xenial_debug]: DL3008 requires to specify version. Don't know how to specify version for `libstdc++6`

File: docker/Dockerfile.xenial_debug (hunk 4 above)
Comment (crazyxy) [reply docker/Dockerfile.xenial_debug]: Yes. I don't have strong preference.

File: docker/Dockerfile.xenial_debug (hunk 2 above)
Comment (PiotrSikora) [reply docker/Dockerfile.xenial_debug]: @howardjohn that's different library (libc++ vs libstdc++).

File: docker/Dockerfile.xenial_debug (hunk 4 above)
Comment (PiotrSikora) [reply docker/Dockerfile.xenial_debug]: Please don't, because then, in theory, `ppa:ubuntu-toolchain-r/test` could be used for other packages.

Comment (PiotrSikora) [reply docker/Dockerfile.xenial_debug]: i.e. do it after `apt-get upgrade`

**Comment by crazyxy** [comment]
Close this pr as @PiotrSikora is working to build the Envoy statically.

**Pull request closed without merging**